src/libmeasurement_kit/ooni/utils.cpp

```diff
@@ -23,141 +23,103 @@ void resolver_lookup(Callback<Error, std::string> callback, Settings settings,
+        if (result == nullptr) {
+            return GeoipCountryCodeLookupError();
         }
-    }
-    GeoIPLookup gl;
-    memset(&gl, 0, sizeof(gl));
-
-    const char *result;
-    result = GeoIP_country_name_by_name_gl(gi_country, ip.c_str(), &gl);
-    if (result == nullptr) {
-        return GenericError();
-    }
-    std::string country_name = result;
-    return country_name;
+        std::string country_code = result;
+        return country_code;
+    }, logger);
 }
 
-ErrorOr<std::string> IPLocation::resolve_city_name(std::string ip,
-                                                   Var<Logger> logger) {
-    if (gi_city == nullptr) {
-        gi_city = GeoIP_open(path_city.c_str(), GEOIP_MEMORY_CACHE);
-        if (gi_country == nullptr) {
-            logger->warn("IPLocation: cannot open geoip city database");
-            return CannotOpenGeoIpCityDatabaseError();
+ErrorOr<std::string> GeoipDatabase::resolve_country_name(
+            std::string ip, Var<Logger> logger) {
+    return with_open_database_do([=]() -> ErrorOr<std::string> {
+        GeoIPLookup gl;
+        memset(&gl, 0, sizeof(gl));
+        const char *result;
+        result = GeoIP_country_name_by_name_gl(db.get(), ip.c_str(), &gl);
+        if (result == nullptr) {
+            return GeoipCountryNameLookupError();
         }
-    }
-    GeoIPRecord *gir = GeoIP_record_by_name(gi_city, ip.c_str());
-    if (gir == nullptr) {
-        return GenericError();
-    }
-    std::string result;
-    if (gir->city != nullptr) {
-        result = gir->city;
-    }
-    GeoIPRecord_delete(gir);
-    return result;
+        std::string country_name = result;
+        return country_name;
+    }, logger);
 }
 
-ErrorOr<std::string> IPLocation::resolve_asn(std::string ip,
-                                             Var<Logger> logger) {
-    if (gi_asn == nullptr) {
-        gi_asn = GeoIP_open(path_asn.c_str(), GEOIP_MEMORY_CACHE);
-        if (gi_asn == nullptr) {
-            logger->warn("IPLocation: cannot open geoip asn database");
-            return CannotOpenGeoIpAsnDatabaseError();
+ErrorOr<std::string> GeoipDatabase::resolve_city_name(
+            std::string ip, Var<Logger> logger) {
+    return with_open_database_do([=]() -> ErrorOr<std::string> {
+        GeoIPRecord *gir = GeoIP_record_by_name(db.get(), ip.c_str());
+        if (gir == nullptr) {
+            return GeoipCityLookupError();
         }
-    }
-    GeoIPLookup gl;
-    memset(&gl, 0, sizeof(gl));
-
-    char *res;
-    res = GeoIP_name_by_name_gl(gi_asn, ip.c_str(), &gl);
-    if (res == nullptr) {
-        return GenericError();
-    }
-    std::string asn = res;
-    asn = split(asn).front(); // We only want ASXX
-    free(res);
-    return asn;
+        std::string result;
+        if (gir->city != nullptr) {
+            result = gir->city;
+        }
+        GeoIPRecord_delete(gir);
+        return result;
+    }, logger);
```

Comment:
Most of the diff related to this function above disappears if we don't consider space changes